File: backend/schemas/analysis.py
from pydantic import BaseModel, HttpUrl
from typing import Dict, List, Optional


class AnalyzeRequest(BaseModel):
    repo_url: HttpUrl


# ---------------------------------------------------------------------------
# 1. Complexity (Cyclomatic / McCabe)
# ---------------------------------------------------------------------------

class FunctionComplexity(BaseModel):
    name: str
    complexity: int
    lineno: int
    is_high_risk: bool


class FileComplexity(BaseModel):
    file_path: str
    functions: List[FunctionComplexity]
    average_complexity: float
    total_complexity: int
    max_complexity: int
    function_count: int


class HighRiskFunction(BaseModel):
    name: str
    complexity: int
    lineno: int
    file_path: str


class ComplexityReport(BaseModel):
    files: List[FileComplexity]
    average_complexity: float
    total_complexity: int
    function_count: int
    high_risk_functions: List[HighRiskFunction]
    distribution: Dict[str, int]


# ---------------------------------------------------------------------------
# 2. Duplication (Winnowing)
# ---------------------------------------------------------------------------

class DuplicateBlock(BaseModel):
    start: int
    end: int
    code: str


class DuplicatePair(BaseModel):
    file_a: str
    file_b: str
    similarity: float
    blocks_a: List[DuplicateBlock]
    blocks_b: List[DuplicateBlock]
    shared_fingerprints: int


class DuplicationReport(BaseModel):
    duplicate_pairs: List[DuplicatePair]
    duplication_percentage: float
    duplicated_lines: int
    total_lines: int
    pair_count: int


# ---------------------------------------------------------------------------
# 3. Maintainability (Halstead + MI)
# ---------------------------------------------------------------------------

class HalsteadMetrics(BaseModel):
    distinct_operators: int
    distinct_operands: int
    total_operators: int
    total_operands: int
    vocabulary: int
    length: int
    volume: float
    difficulty: float
    effort: float


class FileMaintainability(BaseModel):
    file_path: str
    halstead: HalsteadMetrics
    cyclomatic_complexity: int
    loc: int
    maintainability_index: float
    rating: str


class LowestMaintainability(BaseModel):
    file_path: str
    maintainability_index: float
    rating: str


class MaintainabilityReport(BaseModel):
    files: List[FileMaintainability]
    average_maintainability: float
    rating: str
    lowest_files: List[LowestMaintainability]


# ---------------------------------------------------------------------------
# 4. Security (AST patterns + secrets + dependency CVEs)
# ---------------------------------------------------------------------------

class SecurityIssue(BaseModel):
    line: int
    rule: str
    severity: str
    title: str
    message: str
    recommendation: str
    cwe: Optional[str] = None
    match: Optional[str] = None  # redacted secret, when the rule found one
    in_test: bool = False        # found in test code, so severity was lowered


class FileSecurityIssues(BaseModel):
    file_path: str
    issues: List[SecurityIssue]
    issue_count: int
    highest_severity: str
    is_test: bool = False


class DependencyVulnerability(BaseModel):
    package: str
    version: str
    pinned: bool
    constraint: str
    source: str
    vulnerability_id: str
    cve: Optional[str] = None
    severity: str
    summary: str
    fixed_versions: List[str]
    reference: str


class DependencyScan(BaseModel):
    available: bool
    reason: Optional[str] = None
    dependencies_checked: int
    unpinned_count: int = 0
    vulnerabilities: List[DependencyVulnerability]
    vulnerable_package_count: int


class SecurityReport(BaseModel):
    files: List[FileSecurityIssues]
    dependencies: DependencyScan
    severity_counts: Dict[str, int]
    total_issues: int
    test_issues: int = 0
    affected_files: int
    security_score: float


# ---------------------------------------------------------------------------
# 5. Dead code (reachability analysis)
# ---------------------------------------------------------------------------

class DeadDefinition(BaseModel):
    file_path: str
    name: str
    kind: str          # function | method | class
    lineno: int
    lines: int
    owner: Optional[str] = None
    confidence: str    # high | medium | low


class UnusedImport(BaseModel):
    file_path: str
    lineno: int
    name: str
    statement: str


class UnusedLocal(BaseModel):
    file_path: str
    function: str
    name: str
    lineno: int


class UnreachableCode(BaseModel):
    file_path: str
    lineno: int
    after: str
    after_line: int
    statements: int


class DeadCodeReport(BaseModel):
    dead_functions: List[DeadDefinition]
    dead_classes: List[DeadDefinition]
    unused_imports: List[UnusedImport]
    unused_locals: List[UnusedLocal]
    unreachable_code: List[UnreachableCode]
    counts: Dict[str, int]
    total_items: int
    dead_lines: int
    dead_code_percentage: float
    high_confidence_count: int


# ---------------------------------------------------------------------------
# 6. Type hint coverage
# ---------------------------------------------------------------------------

class FunctionTypeHints(BaseModel):
    name: str
    lineno: int
    parameters: int
    annotated_parameters: int
    has_return_annotation: bool
    coverage: float
    is_public: bool


class FileTypeHints(BaseModel):
    file_path: str
    functions: List[FunctionTypeHints]
    function_count: int
    annotated_slots: int
    total_slots: int
    coverage: float
    fully_typed: int
    partially_typed: int
    untyped: int
    annotated_variables: int
    rating: str


class LowestTypeCoverage(BaseModel):
    file_path: str
    coverage: float
    function_count: int
    untyped: int
    rating: str


class UntypedFunction(BaseModel):
    file_path: str
    name: str
    lineno: int
    parameters: int


class TypeHintReport(BaseModel):
    files: List[FileTypeHints]
    coverage: float
    rating: str
    annotated_slots: int
    total_slots: int
    function_count: int
    fully_typed: int
    partially_typed: int
    untyped: int
    annotated_variables: int
    lowest_files: List[LowestTypeCoverage]
    untyped_public_functions: List[UntypedFunction]


# ---------------------------------------------------------------------------
# 7. Behavioural history (churn, hotspots, coupling, bus factor)
# ---------------------------------------------------------------------------

class ChurnFile(BaseModel):
    file_path: str
    commits: int
    insertions: int
    deletions: int
    churn: int
    weighted_churn: float
    recent_commits: int
    author_count: int
    first_modified: str
    last_modified: str


class Hotspot(BaseModel):
    file_path: str
    risk_score: float
    category: str      # critical | high | moderate | low
    complexity: int
    churn: int
    commits: int
    recent_commits: int
    author_count: int
    primary_author: str
    primary_author_share: float
    last_modified: str


class CouplingPair(BaseModel):
    file_a: str
    file_b: str
    co_changes: int
    commits_a: int
    commits_b: int
    degree: float
    jaccard: float


class Contributor(BaseModel):
    author: str
    commits: int
    lines: int
    share: float
    last_active: str


class AtRiskFile(BaseModel):
    file_path: str
    primary_author: str
    primary_author_share: float
    author_count: int
    commits: int
    churn: int
    bus_factor: int


class BusFactorReport(BaseModel):
    repository_bus_factor: int
    contributor_count: int
    top_contributors: List[Contributor]
    at_risk_files: List[AtRiskFile]
    at_risk_count: int


class HistorySummary(BaseModel):
    critical_hotspots: int
    high_hotspots: int
    coupled_pairs: int
    knowledge_risk_files: int


class HistoryReport(BaseModel):
    available: bool
    reason: Optional[str] = None
    commits_analyzed: int
    contributor_count: int
    period_days: int
    first_commit: Optional[str] = None
    last_commit: Optional[str] = None
    churn_files: List[ChurnFile]
    hotspots: List[Hotspot]
    coupling: List[CouplingPair]
    bus_factor: BusFactorReport
    summary: HistorySummary


# ---------------------------------------------------------------------------
# 8. Trends and regression diffs
# ---------------------------------------------------------------------------

class MetricChange(BaseModel):
    metric: str
    label: str
    unit: str
    before: float
    after: float
    delta: float
    percent_change: Optional[float] = None
    direction: str     # improved | regressed | unchanged | changed
    headline: bool


class ComparisonReport(BaseModel):
    available: bool
    verdict: str       # baseline | improved | regressed | unchanged
    reason: Optional[str] = None
    changes: List[MetricChange]
    regressions: List[MetricChange]
    improvements: List[MetricChange]


class MetricDefinition(BaseModel):
    key: str
    label: str
    unit: str
    higher_is_better: Optional[bool] = None
    headline: bool


class TrendPoint(BaseModel):
    analysis_id: int
    date: str
    commit_sha: Optional[str] = None
    health_score: Optional[float] = None
    average_maintainability: Optional[float] = None
    average_complexity: Optional[float] = None
    duplication_percentage: Optional[float] = None
    security_score: Optional[float] = None
    type_hint_coverage: Optional[float] = None
    high_risk_functions: Optional[float] = None
    critical_security_issues: Optional[float] = None
    dead_code_items: Optional[float] = None
    critical_hotspots: Optional[float] = None
    lines_of_code: Optional[float] = None
    python_files: Optional[float] = None


class TrendSeries(BaseModel):
    repo_name: str
    points: List[TrendPoint]
    run_count: int
    metrics: List[MetricDefinition]
    latest_comparison: Optional[ComparisonReport] = None


class AnalysisRun(BaseModel):
    id: int
    repo_name: str
    repo_url: str
    status: str
    health_score: Optional[int] = None
    commit_sha: Optional[str] = None
    ref: Optional[str] = None
    created_at: Optional[str] = None


# ---------------------------------------------------------------------------
# Auxiliary: bad practices
# ---------------------------------------------------------------------------

class BadPractice(BaseModel):
    line: int
    type: str
    message: str


class FileBadPractices(BaseModel):
    file_path: str
    issues: List[BadPractice]


# ---------------------------------------------------------------------------
# Dashboard summary + health score
# ---------------------------------------------------------------------------

class HealthScore(BaseModel):
    score: int
    grade: str
    components: Dict[str, int]
    weights: Dict[str, float]


class RepositorySummary(BaseModel):
    repository_name: str
    python_files: int
    total_functions: int
    total_classes: int
    lines_of_code: int
    average_complexity: float
    duplication_percentage: float
    average_maintainability: float
    security_score: float
    type_hint_coverage: float
    dead_code_items: int
    health_score: HealthScore


# ---------------------------------------------------------------------------
# top-level response
# ---------------------------------------------------------------------------

class AnalyzeResponse(BaseModel):
    summary: RepositorySummary
    commit_sha: Optional[str] = None
    complexity: ComplexityReport
    duplication: DuplicationReport
    maintainability: MaintainabilityReport
    security: SecurityReport
    dead_code: DeadCodeReport
    type_hints: TypeHintReport
    history: HistoryReport
    bad_practices: List[FileBadPractices]
    comparison: Optional[ComparisonReport] = None

File: frontend/app/types/analysis.ts
// ---- 1. Complexity ----
export interface FunctionComplexity {
  name: string;
  complexity: number;
  lineno: number;
  is_high_risk: boolean;
}
export interface FileComplexity {
  file_path: string;
  functions: FunctionComplexity[];
  average_complexity: number;
  total_complexity: number;
  max_complexity: number;
  function_count: number;
}
export interface HighRiskFunction {
  name: string;
  complexity: number;
  lineno: number;
  file_path: string;
}
export interface ComplexityReport {
  files: FileComplexity[];
  average_complexity: number;
  total_complexity: number;
  function_count: number;
  high_risk_functions: HighRiskFunction[];
  distribution: { low: number; moderate: number; high: number };
}

// ---- 2. Duplication (Winnowing) ----
export interface DuplicateBlock {
  start: number;
  end: number;
  code: string;
}
export interface DuplicatePair {
  file_a: string;
  file_b: string;
  similarity: number;
  blocks_a: DuplicateBlock[];
  blocks_b: DuplicateBlock[];
  shared_fingerprints: number;
}
export interface DuplicationReport {
  duplicate_pairs: DuplicatePair[];
  duplication_percentage: number;
  duplicated_lines: number;
  total_lines: number;
  pair_count: number;
}

// ---- 3. Maintainability (Halstead + MI) ----
export interface HalsteadMetrics {
  distinct_operators: number;
  distinct_operands: number;
  total_operators: number;
  total_operands: number;
  vocabulary: number;
  length: number;
  volume: number;
  difficulty: number;
  effort: number;
}
export interface FileMaintainability {
  file_path: string;
  halstead: HalsteadMetrics;
  cyclomatic_complexity: number;
  loc: number;
  maintainability_index: number;
  rating: string;
}
export interface LowestMaintainability {
  file_path: string;
  maintainability_index: number;
  rating: string;
}
export interface MaintainabilityReport {
  files: FileMaintainability[];
  average_maintainability: number;
  rating: string;
  lowest_files: LowestMaintainability[];
}

// ---- 4. Security ----
export type Severity = "critical" | "high" | "medium" | "low";

export interface SecurityIssue {
  line: number;
  rule: string;
  severity: Severity;
  title: string;
  message: string;
  recommendation: string;
  cwe: string | null;
  match?: string | null;
  in_test: boolean;
}
export interface FileSecurityIssues {
  file_path: string;
  issues: SecurityIssue[];
  issue_count: number;
  highest_severity: Severity;
  is_test: boolean;
}
export interface DependencyVulnerability {
  package: string;
  version: string;
  pinned: boolean;
  constraint: string;
  source: string;
  vulnerability_id: string;
  cve: string | null;
  severity: Severity;
  summary: string;
  fixed_versions: string[];
  reference: string;
}
export interface DependencyScan {
  available: boolean;
  reason: string | null;
  dependencies_checked: number;
  unpinned_count: number;
  vulnerabilities: DependencyVulnerability[];
  vulnerable_package_count: number;
}
export interface SecurityReport {
  files: FileSecurityIssues[];
  dependencies: DependencyScan;
  severity_counts: Record<Severity, number>;
  total_issues: number;
  test_issues: number;
  affected_files: number;
  security_score: number;
}

// ---- 5. Dead code ----
export interface DeadDefinition {
  file_path: string;
  name: string;
  kind: string;
  lineno: number;
  lines: number;
  owner: string | null;
  confidence: "high" | "medium" | "low";
}
export interface UnusedImport { file_path: string; lineno: number; name: string; statement: string; }
export interface UnusedLocal { file_path: string; function: string; name: string; lineno: number; }
export interface UnreachableCode {
  file_path: string;
  lineno: number;
  after: string;
  after_line: number;
  statements: number;
}
export interface DeadCodeReport {
  dead_functions: DeadDefinition[];
  dead_classes: DeadDefinition[];
  unused_imports: UnusedImport[];
  unused_locals: UnusedLocal[];
  unreachable_code: UnreachableCode[];
  counts: Record<string, number>;
  total_items: number;
  dead_lines: number;
  dead_code_percentage: number;
  high_confidence_count: number;
}

// ---- 6. Type hint coverage ----
export interface FunctionTypeHints {
  name: string;
  lineno: number;
  parameters: number;
  annotated_parameters: number;
  has_return_annotation: boolean;
  coverage: number;
  is_public: boolean;
}
export interface FileTypeHints {
  file_path: string;
  functions: FunctionTypeHints[];
  function_count: number;
  annotated_slots: number;
  total_slots: number;
  coverage: number;
  fully_typed: number;
  partially_typed: number;
  untyped: number;
  annotated_variables: number;
  rating: string;
}
export interface LowestTypeCoverage {
  file_path: string;
  coverage: number;
  function_count: number;
  untyped: number;
  rating: string;
}
export interface UntypedFunction { file_path: string; name: string; lineno: number; parameters: number; }
export interface TypeHintReport {
  files: FileTypeHints[];
  coverage: number;
  rating: string;
  annotated_slots: number;
  total_slots: number;
  function_count: number;
  fully_typed: number;
  partially_typed: number;
  untyped: number;
  annotated_variables: number;
  lowest_files: LowestTypeCoverage[];
  untyped_public_functions: UntypedFunction[];
}

// ---- 7. Behavioural history (churn, hotspots, coupling, bus factor) ----
export interface ChurnFile {
  file_path: string;
  commits: number;
  insertions: number;
  deletions: number;
  churn: number;
  weighted_churn: number;
  recent_commits: number;
  author_count: number;
  first_modified: string;
  last_modified: string;
}
export interface Hotspot {
  file_path: string;
  risk_score: number;
  category: "critical" | "high" | "moderate" | "low";
  complexity: number;
  churn: number;
  commits: number;
  recent_commits: number;
  author_count: number;
  primary_author: string;
  primary_author_share: number;
  last_modified: string;
}
export interface CouplingPair {
  file_a: string;
  file_b: string;
  co_changes: number;
  commits_a: number;
  commits_b: number;
  degree: number;
  jaccard: number;
}
export interface Contributor { author: string; commits: number; lines: number; share: number; last_active: string; }
export interface AtRiskFile {
  file_path: string;
  primary_author: string;
  primary_author_share: number;
  author_count: number;
  commits: number;
  churn: number;
  bus_factor: number;
}
export interface BusFactorReport {
  repository_bus_factor: number;
  contributor_count: number;
  top_contributors: Contributor[];
  at_risk_files: AtRiskFile[];
  at_risk_count: number;
}
export interface HistoryReport {
  available: boolean;
  reason: string | null;
  commits_analyzed: number;
  contributor_count: number;
  period_days: number;
  first_commit: string | null;
  last_commit: string | null;
  churn_files: ChurnFile[];
  hotspots: Hotspot[];
  coupling: CouplingPair[];
  bus_factor: BusFactorReport;
  summary: {
    critical_hotspots: number;
    high_hotspots: number;
    coupled_pairs: number;
    knowledge_risk_files: number;
  };
}

// ---- 8. Trends and regression diffs ----
export type ChangeDirection = "improved" | "regressed" | "unchanged" | "changed";

export interface MetricChange {
  metric: string;
  label: string;
  unit: string;
  before: number;
  after: number;
  delta: number;
  percent_change: number | null;
  direction: ChangeDirection;
  headline: boolean;
}
export interface ComparisonReport {
  available: boolean;
  verdict: "baseline" | "improved" | "regressed" | "unchanged";
  reason: string | null;
  changes: MetricChange[];
  regressions: MetricChange[];
  improvements: MetricChange[];
}
export interface MetricDefinition {
  key: string;
  label: string;
  unit: string;
  higher_is_better: boolean | null;
  headline: boolean;
}
export interface TrendPoint {
  analysis_id: number;
  date: string;
  commit_sha: string | null;
  [metric: string]: number | string | null;
}
export interface TrendSeries {
  repo_name: string;
  points: TrendPoint[];
  run_count: number;
  metrics: MetricDefinition[];
  latest_comparison: ComparisonReport | null;
}

// ---- Auxiliary: bad practices ----
export interface BadPractice { line: number; type: string; message: string; }
export interface FileBadPractices { file_path: string; issues: BadPractice[]; }

// ---- Dashboard summary + health ----
export interface HealthScore {
  score: number;
  grade: string;
  components: { maintainability: number; complexity: number; duplication: number };
  weights: { maintainability: number; complexity: number; duplication: number };
}
export interface RepositorySummary {
  repository_name: string;
  python_files: number;
  total_functions: number;
  total_classes: number;
  lines_of_code: number;
  average_complexity: number;
  duplication_percentage: number;
  average_maintainability: number;
  security_score: number;
  type_hint_coverage: number;
  dead_code_items: number;
  health_score: HealthScore;
}

// ---- top-level ----
export interface AnalysisReport {
  summary: RepositorySummary;
  commit_sha: string | null;
  complexity: ComplexityReport;
  duplication: DuplicationReport;
  maintainability: MaintainabilityReport;
  security: SecurityReport;
  dead_code: DeadCodeReport;
  type_hints: TypeHintReport;
  history: HistoryReport;
  bad_practices: FileBadPractices[];
  comparison: ComparisonReport | null;
}
